### PR TITLE
[wip] Drivetrain Functionality

### PR DESCRIPTION
**Pull request by brownkat6**

### CODE REVIEW

File: src/main/java/frc/robot/OI.java [hunk 1]
```diff
@@ -15,38 +15,26 @@
...
-    // button.whenReleased(new ExampleCommand());
     private static OI INSTANCE = new OI();
     // declare buttons, joysticks here
     private Joystick leftFlight;
     private Joystick rightFlight;
 
+    public double removeDeadband(double y){
+        if (Math.abs(y)<=.05){
+            return 0;
+        }else{
```
Comment (c-x-berger) [inline src/main/java/frc/robot/OI.java]: Can you do Ctrl+Alt+L on this file? Consistent formatting makes code much easier to read.

File: src/main/java/frc/robot/RobotMap.java [hunk 2]
```diff
@@ -23,6 +23,20 @@
     // number and the module. For example you with a rangefinder:
     // public static final int rangefinderPort = 1;
     // public static final int rangefinderModule = 1;
+
+    public class PDP{
```
Comment (c-x-berger) [inline src/main/java/frc/robot/RobotMap.java]: Again, Ctrl+Alt+L. Also, good idea having constants organized into subclasses.

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 3]
```diff
@@ -17,13 +22,49 @@ private Drivetrain() {
      * @param leftSpeed  Speed of left side.
      * @param rightSpeed Speed of right side.
      */
-    public void tankDrive(double leftSpeed, double rightSpeed) {
+    public void tankDrive(double leftPower, double rightPower)
+    {
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: Formatting, formatting, formatting. `{` goes on the same line as the method declaration.

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 4]
```diff
@@ -17,13 +22,49 @@ private Drivetrain() {
      * @param leftSpeed  Speed of left side.
      * @param rightSpeed Speed of right side.
      */
-    public void tankDrive(double leftSpeed, double rightSpeed) {
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: Please leave these as `leftSpeed` and `rightSpeed` or change the JavaDoc (comment) accordingly.

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 5]
```diff
@@ -17,13 +22,49 @@ private Drivetrain() {
...
+
+        if(!checkTankDriveNotBroken(leftPower,rightPower)){
+            this.stop();
+        }
     }
 
-    public static Drivetrain getInstance() {
-        return INSTANCE;
+    //this checks to make sure that if the robot is told to go forward, its actually traveling forward instead of backward
+    private boolean checkTankDriveNotBroken(double leftPower, double rightPower){
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: This could use a more descriptive name, as well as proper JavaDoc (see `tankDrive` for an example.)

File: src/main/java/frc/robot/subsystems/PDP.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package frc.robot.subsystems;
+
+
+import edu.wpi.first.wpilibj.PowerDistributionPanel;
+import edu.wpi.first.wpilibj.command.Subsystem;
+import frc.robot.RobotMap;
+
+public class PDP extends Subsystem {
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/PDP.java]: I'm not sure how I feel about having things like the PDP (which to me, seem like sensors) as full-fledged `Subsystem`s. WPI describes `Subsystem`s as ["parts of your robot that are independently controlled"](https://wpilib.screenstepslive.com/s/currentCS/m/java/l/599735-simple-subsystems), which to me means this isn't the intended use case. The PDP isn't "independently controlled" - it's not really controlled at all.

That said, the PDP doesn't clearly "belong" to any system on the robot (like Talons etc.), so... I dunno. I'll give it more thought.

File: src/main/java/frc/robot/OI.java [hunk 7]
```diff
@@ -15,38 +15,26 @@
...
+            return y;
+        }
+    }
+    public double getLeftY(){
```
Comment (c-x-berger) [inline src/main/java/frc/robot/OI.java]: [Redefining these breaks the build.](https://travis-ci.org/BHSSFRC/3494_2019_repo/builds/478553551)

**Review by c-x-berger — CHANGES_REQUESTED** [review]
See comments. In general, you need to do Ctrl+Alt+L on everything. (You can actually do this by clicking on `src` in the Project panel on the left and running the key chord.)

File: src/main/java/frc/robot/OI.java [hunk 8]
```diff
@@ -15,62 +15,33 @@
...
-        rightFlight = new Joystick(RobotMap.RIGHT_JOY);
+    public double removeDeadband(double y){
+        if (Math.abs(y)<=.05){
+            return 0;
+        }else{
+            return y;
+        }
     }
-
-    /**
```
Comment (c-x-berger) [inline src/main/java/frc/robot/OI.java]: Leave these comments in, please!

**Review by c-x-berger — CHANGES_REQUESTED** [review]
Re-add documentation comments, please!

File: gradle/wrapper/gradle-wrapper.properties [hunk 9]
```diff
@@ -1,5 +1,6 @@
+#Fri Jan 11 17:16:32 EST 2019
 distributionBase=GRADLE_USER_HOME
 distributionPath=permwrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.0-bin.zip
 zipStoreBase=GRADLE_USER_HOME
-zipStorePath=permwrapper/dists
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.0-all.zip
```
Comment (c-x-berger) [inline gradle/wrapper/gradle-wrapper.properties]: Why did you change this file?

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 10]
```diff
@@ -17,13 +24,56 @@ private Drivetrain() {
...
+
+        if (Math.abs(rightCounter)!=3 || Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.rightMasterChannel)) != Math.signum(leftPower)){
+            System.out.println("One of the left motors is jammed! RIP");
+            return false;
+        }
+        return true;
+    }
+
+    public void stop()
+    {
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: How many times am I gonna say formatting before you do it?

File: src/main/java/frc/robot/subsystems/NavX.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+import com.kauailabs.navx.frc.AHRS;
+import edu.wpi.first.wpilibj.SerialPort;
+import edu.wpi.first.wpilibj.command.Subsystem;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.robot.RobotMap;
+
+public class NavX extends Subsystem {
+
+    private NavX INSTANCE = NavX();
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/NavX.java]: `= new NavX()` or it doesn't build :stuck_out_tongue:

**Review by c-x-berger — CHANGES_REQUESTED** [review]
Your code still needs formatting, and ~~has~~ had syntax errors besides.

**Comment by brownkat6** [comment]
private static OI INSTANCE = new OI();
     // declare buttons, joysticks here
     private Joystick leftFlight;
     private Joystick rightFlight;

+    public double removeDeadband(double y){
+        if (Math.abs(y)<=.05){
+            return 0;
+        }else{
+            return y;
+        }
+    }
+    public double getLeftY(){

What is being redefined?

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 12]
```diff
@@ -18,12 +25,64 @@ private Drivetrain() {
...
+    }
+
+    public void stop() {
+        this.tankDrive(0, 0);
     }
 
+    @Override
+    public void periodic() {
+        if (SmartDashboard.getBoolean("Display Drivetrain data?", false)) {
+            System.out.println("The left side: " + PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.leftMasterChannel) + ", "
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:75]: Using `println` for debugging sounds like a good way to fill up the console to unreadability instantly (and if it gets left on in comp, I can attest to it having eaten mountains of bandwidth somehow...)

@BoxWizard000000, was this your idea?

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 13]
```diff
@@ -18,12 +25,64 @@ private Drivetrain() {
...
+        }
     }
 
-    public static Drivetrain getInstance() {
-        return INSTANCE;
+    /**
+     * checkDriveMotorsGoCorrectDirection sanity check
+     * this checks to make sure that if the robot is told to go forward, its actually traveling forward instead of backward
+     * @param leftSpeed  Speed of left side.
+     * @param rightSpeed Speed of right side.
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: In general, `@param` statements go at the end of these kinds of comments. IntelliJ formatting should catch that (though it's not perfect.)

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 14]
```diff
@@ -1,13 +1,20 @@
...
+import frc.robot.RobotMap;
 import frc.robot.commands.drive.Drive;
 
 public class Drivetrain extends Subsystem {
     // Put methods for controlling this subsystem
     // here. Call these from Commands.
     private static Drivetrain INSTANCE = new Drivetrain();
 
+    public static Drivetrain getInstance() {
+        return INSTANCE;
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: I didn't point it out before, but you probably shouldn't reorder methods like this. If nothing else, it clutters up diffs a bit. Strictly speaking, it's not _wrong_, but it makes this review process harder.

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 15]
```diff
@@ -18,12 +25,64 @@ private Drivetrain() {
...
+     *                   and that they are traveling the same direction as the variable leftSpeed is telling them to go
+     * Ditto for right side.
+     */
+    private boolean checkDriveMotorsGoCorrectDirection(double leftSpeed, double rightSpeed) {
+        int leftCounter = 0;
+        leftCounter += Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.leftMasterChannel));
+        leftCounter += Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.leftFollower1Channel));
+        leftCounter += Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.leftFollower2Channel));
+
+        if (Math.abs(leftCounter) != 3 || Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.leftMasterChannel)) != Math.signum(leftSpeed)) {
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:51]: I'm not thoroughly convinced that this is be best way to detect that the motors are going the right way. Give me some more time to think it over.

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 16]
```diff
@@ -18,12 +25,64 @@ private Drivetrain() {
...
+            System.out.println("One of the left motors is jammed! RIP");
+            return false;
+        }
+
+        int rightCounter = 0;
+        rightCounter += Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.rightMasterChannel));
+        rightCounter += Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.rightFollower1Channel));
+        rightCounter += Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.rightFollower2Channel));
+
+        if (Math.abs(rightCounter) != 3 || Math.signum(PDP.getInstance().getCurrent(RobotMap.DRIVETRAIN.rightMasterChannel)) != Math.signum(leftSpeed)) {
```
Comment (c-x-berger) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: Comparing to the wrong variable (`Math.signum(leftSpeed)` != `Math.signum(rightSpeed)`)

File: src/main/java/frc/robot/RobotMap.java [hunk 17]
```diff
@@ -23,6 +23,20 @@
     // number and the module. For example you with a rangefinder:
     // public static final int rangefinderPort = 1;
     // public static final int rangefinderModule = 1;
+
+    public class PDP {
+        public static final int CAN_ID = 0;
+    }
```
Comment (c-x-berger) [inline src/main/java/frc/robot/RobotMap.java]: ... what?

The default constructor for a PDP uses CAN ID 0 (and we don't need to change it since we only have one PDP.) Why did you decide to explicitly specify that anyway?

**Review by c-x-berger — CHANGES_REQUESTED** [review]
Looks mostly good, would prefer that you left methods in original order (see `Drivetrain` class, as well as `OI`.)

`checkDriveMotorsGoCorrectDirection` is... a less than ideal name, still, but technically more descriptive I suppose. Speaking of that sanity check, it's got a massive breaking flaw. You should see it pretty quick.

File: src/main/java/frc/robot/subsystems/Drivetrain.java (hunk 14 above)
Comment (c-x-berger) [reply src/main/java/frc/robot/subsystems/Drivetrain.java]: (There's the same thing going on in `OI`, I think.)

File: src/main/java/frc/robot/subsystems/Drivetrain.java (hunk 15 above)
Comment (c-x-berger) [reply src/main/java/frc/robot/subsystems/Drivetrain.java:51]: So I did sleep on it, and I'm still looking at this thinking that there might be a situation where the motors are drawing "negative current" according the the PDP when a positive signal is sent, but that remains desired operation. I'm going to have to test this on a practice base though. If that is the case, we have to do some changes to this statement (because I doubt @Feith27 will be likely to hardware invert motors.)

If nothing else, this is definitely something that I want to see "battle tested" on real robots.

**Review by c-x-berger — APPROVED** [review]

**Comment by c-x-berger** [comment]
Looks mergable to me. I still have my issues, but they're either too nitpicky or too grand (e.g. sensor subsystems) to bother fixing before merge.

A good compromise on the sensors issue might be a grand unified "Sensors" susbsytem (GRUSS lol), however, that would probably lock up commands a fair amount (if command A only needs the NavX and command B only needs the PDP they'll still block each other.) And you still have subsystems using each other, which seems to defeat some of their purpose.

A case can _definitely_ be made for saying that the NavX belongs to the drivetrain, I know that much.

File: src/main/java/frc/robot/subsystems/PDP.java (hunk 6 above)
Comment (c-x-berger) [reply src/main/java/frc/robot/subsystems/PDP.java]: Gave it thought! Make a general `Sensors` class that doesn't extend `Subsystem` with your utility methods for each sensor in it. Make it a singleton if you want as well.
